src/main/resources/rs117/hd/frag.glsl

```diff
@@ -152,24 +153,39 @@ void main() {
         float selfShadowing = 0;
         vec3 fragPos = IN.position;
         #if PARALLAX_OCCLUSION_MAPPING
-        mat3 invTBN = inverse(TBN);
-        vec3 tsViewDir = invTBN * viewDir;
-        vec3 tsLightDir = invTBN * -lightDir;
+        if(material1.displacementMap != -1 || material2.displacementMap != -1 || material3.displacementMap != -1)
+        {
+            const float minLayers = 1;
+            const float maxLayers = 16;
+
+            float layerDistFade = 1.0 - clamp((length(viewVec) - 500.0) / 2000.0, 0.0, 1.0);
+            if(layerDistFade > 0.0) {
+                mat3 invTBN = inverse(TBN);
```

Comment:
I think we can do `transpose(TBN)` here instead of `inverse(TBN)` to save a tiny bit of computation also, since it *should* be an orthogonal matrix :slightly_smiling_face:

Reply:
Ah, I guess I never got the TBN quite right then :sweat_smile: Maybe worth revisiting at some point.